### ISSUE TITLE
Possible issues with the function that defines grammar for capturing Tool Call.

### ISSUE THREAD

**Issue by anshmn** [guidance-ai/guidance#602]
**The bug**
The function that defines the Grammar for Triggering the tool Usage seems to behave in an un-expected manner.

Consider the following Code  , The definition for expression() is same as in Readme .
```python
from guidance import capture, Tool
@guidance(stateless=True)
def calculator_call(lm):
    # capture just 'names' the expression, to be saved in the LM state
    return lm + expression()
@guidance
def calculator(lm):
    lm += f'$$$$'
    return lm
calculator_tool = Tool(calculator_call(), calculator)
lm5 =  models.Mock(b"<s> 5+4*5  Some English Text <s>")
lm5 += gen(max_tokens=30, tools=[calculator_tool], stop='\n\n')
```

Output :   5+4*5  +9$$$$ewxbewsawrglbrsgvqmntnyrfylaxttvaluu,giicndoeqgspkjloukxo(1$$$$

Rather than Calling the calculator after encountering the Token Sequence "5+4*5" , it seems that the function calculator_call changes the model state in an odd fashion causing it to generate the token +9 following which the call to calculator is finally made and after which the model goes on to produce the Gibberish text . However Intersetingly , the call to calculator is made correctly later on , when the token corresponding to "1" is generated.

 Also, when I change the definiton of the calculator_call to 

```python
def calculator_call(lm):
    # capture just 'names' the expression, to be saved in the LM state
    return lm + capture(expression(),name = 'A')
```

Output :   5$$$$ln ymoadxdggcjbewxbewsawrglbrsgvqmntnyrfylaxttvaluu,gi6$$$$

So , In this case the call to calculator is made as soon as the number 5 is encountered , i.e it greedily finds the shortest pattern matching the grammar corresponding to expression() and calls the calculator ( I assume this is how it has been implemented ).

**To Reproduce**
Give a full working code snippet that can be pasted into a notebook cell or python file. Make sure to include the LLM load step so we know which model you are using.
```python
import guidance
from guidance import one_or_more, select, zero_or_more
# stateless=True indicates this function does not depend on LLM generations
@guidance(stateless=True)
def number(lm):
    n = one_or_more(select(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']))
    # Allow for negative or positive numbers
    return lm + select(['-' + n, n])

@guidance(stateless=True)
def operator(lm):
    return lm + select(['+' , '*', '**', '/', '-'])

@guidance(stateless=True)
def expression(lm):
    # Either
    # 1. A number (terminal)
    # 2. two expressions with an operator and optional whitespace
    # 3. An expression with parentheses around it
    return lm + select([
        number(),
        expression() + zero_or_more(' ') +  operator() + zero_or_more(' ') +  expression(),
        '(' + expression() + ')'
    ])



from guidance import capture, Tool
@guidance(stateless=True)
def calculator_call(lm):
    # capture just 'names' the expression, to be saved in the LM state
    return lm + (expression())
@guidance
def calculator(lm):
    lm += f'$$$$'
    return lm
calculator_tool = Tool(calculator_call(), calculator)
lm5 =  models.Mock(b"<s> 5+4*5  Some English Text <s>")
lm5 += gen(max_tokens=30, tools=[calculator_tool], stop='\n\n')
```

**Intended Use**
I want to call the calculator as soon as an Expression is encountered ( but not in a greedy fashion , i.e the longest expression should be allowed to generate before making the call ) . 
Also , I don't want to have to guide the LLM explicitly to generate the syntax " calculator( expression()  ) " as is the case in the Readme example provided.

Any help is appreciated.

**System info (please complete the following information):**
 - OS (e.g. Ubuntu, Windows 11, Mac OS, etc.) :  Ubuntu
 - Guidance Version (`guidance.__version__`): 0.1.10

**Comment by slundberg** [comment]
Hi @anshmn ! Good questions. The reason the calculator is not called when you get to `5+4*5` is because the tool call is not greedy matching, it allows the model to make a longer tool call if it wants to (or emit the stop tokens, which you have set to `\n\n`). Since a space is a valid tool call continuation of `5+4*5` you get `5+4*5 ` generated by the model. At this point we are committed to making a tool call because `5+4*5` already happened, but we don't have a complete match yet, so the model is force to finish the grammar, which in this case results in `+9`) once the grammar is complete the stop token is chosen and the call is made.

I am going to leave this issue open however, because the change in greedy vs. non-greedy when you add a capture does seem off. Thanks for flagging that!

Some notes:

As part of debugging this I added in another PR (to be merged soon) a `match` method on the GrammarFunction so you can test matches more easily:
<img width="405" alt="image" src="https://github.com/guidance-ai/guidance/assets/3740613/8fd5c1a2-d906-419c-9e40-bb071d928049">

In your original output `Output : 5+4*5 +9$$$$ewxbewsawrglbrsgvqmntnyrfylaxttvaluu�,giicndoeqgspkjloukxo(1$$$$` two calls are made (each $$$$ is a call result). The random text is just what the Mock model generates when it does not have a matching prefix.

**Comment by anshmn** [comment]
Thanks for the detailed reply @slundberg . 
Their seems to be some other issues with the way tool call is being made .

Consider the following code snippet 
```python
from guidance import capture, Tool
lst=[]
@guidance(stateless=True)
def calculator_call(lm):
    # capture just 'names' the expression, to be saved in the LM state
    return lm + expression()
                 
@guidance
def calculator(lm):
    lst.append(str(lm))
    return lm
    
calculator_tool = Tool(calculator_call(), calculator)
lm5 =  models.Mock(b"<s>5456 +445 * 5Some English Text<s>")
lm5.echo = False
lm5 += gen(max_tokens=30, tools=[calculator_tool], stop='<s>')
print(lst)
```
['5', '54', '545', '5456 +4', '5456 +44', '5456 +445 * 5']

As is evident from the code output above  , the call to calculator is made as soon as a valid expression is encountered provided that the output generation after the point where the call is made still remains a valid expression !!

This explains the call being made at '5', '54', '545' but not '5456' because "+445 * 5" won't remain a valid expression .
